### PR TITLE
[KAFKA-9826] Handle an unaligned first dirty offset during log cleani…

### PR DESCRIPTION
**Pull request by steverod**
…ng. (#8469)

In KAFKA-9826, a log whose first dirty offset was past the start of the active segment and past the last cleaned point resulted in an endless cycle of picking the segment to clean and discarding it. Though this didn't interfere with cleaning other log segments, it kept the log cleaner thread continuously busy (potentially wasting CPU and impacting other running threads) and filled the logs with lots of extraneous messages.

This was determined to be because the active segment was getting mistakenly picked for cleaning, and because the logSegments code handles (start == end) cases only for (start, end) on a segment boundary: the intent is to return a null list, but if they're not on a segment boundary, the routine returns that segment.

This fix has two parts:

It changes logSegments to handle start==end by returning an empty List always.

It changes the definition of calculateCleanableBytes to not consider anything past the UncleanableOffset; previously, it would potentially shift the UncleanableOffset to match the firstDirtyOffset even if the firstDirtyOffset was past the firstUncleanableOffset. This has no real effect now in the context of the fix for (1) but it makes the code read more like the model that the code is attempting to follow.

These changes require modifications to a few test cases that handled this particular test case; they were introduced in the context of KAFKA-8764. Those situations are now handled elsewhere in code, but the tests themselves allowed a DirtyOffset in the active segment, and expected an active segment to be selected for cleaning.

Reviewer: Jun Rao <junrao@gmail.com>

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by junrao** [comment]
ok to test

**Review by junrao — APPROVED** [review]
@steverod : Thanks for the PR. LGTM. Waiting for the tests to pass before merge.

**Comment by junrao** [comment]
retest this please

**Comment by junrao** [comment]
@steverod : Does the JDK 8 and Scala 2.12 tests pass for you locally? Not sure why the jenkins test failed.

**Comment by steverod** [comment]
retest this please

**Comment by steverod** [comment]
@junrao my local runs with `-PscalaVersion=2.12` compile and run fine. I'm not able to access the  check report so I can't see what's going on there.

**Comment by junrao** [comment]
The tests results are too old to access now. Let me run the test again.

**Comment by junrao** [comment]
retest this please

**Comment by junrao** [comment]
ok to test

**Comment by junrao** [comment]
test this please

**Comment by junrao** [comment]
The tests passed. Merging to 2.5.

**Pull request merged**
